### DISCUSSION TITLE
Difference between Modes 1 and 2 in RPU extraction

### DISCUSSION THREAD

**Discussion by markojurkovic** [quietvoid/dovi_tool#117] [Q&A]
Hello quietvoid, first of all, thank you for creating and publishing this tool! Just have a quick question regarding the extraction of RPU from an EL hevc file. What exactly is the difference between extracting Level 8.1 RPU (--mode 1) vs "MEL RPU" (--mode 2)? Can the resulting RPU files be used interchangeably? For example, if I want to create a Profile 4 output (which optionally supports MEL instead of the default FEL), can I use either Mode 1 or 2? Can I use Level 8.1 RPU for other output Levels? Isn't Level 8.1 also "MEL only"?

Thank you in advance for clarification...

**Comment by quietvoid** [comment]
`dovi_tool` does not support profile 4 as an output. It's a deprecated format.

The two modes can only be used interchangeably if the RPU is profile 7 MEL.
Profile 8.1 differs in that it's not meant to be used along an enhancement layer, while profile 7 MEL is.

> Can I use Level 8.1 RPU for other output Levels? 

I'm not sure what you mean. Dolby Vision output is always full range for profiles other than 4.

**Comment by markojurkovic** [comment]
Thank you very much for responding so quickly. So, if I got it right, both MEL and FEL layers on a UHD Blu-ray are composed of a video stream (EL) + RPU? Or is MEL just RPU being added on top of the Base Layer?

**Reply by quietvoid** [reply]
> So, if I got it right, both MEL and FEL layers on a UHD Blu-ray are composed of a video stream (EL) + RPU?

Yes that's correct.

**Comment by markojurkovic** [comment]
OK, great, thanks for confirming!

Final thought: to tie this to my original question and what I wanted to ask, is basically when should we use Mode 1 or Mode 2 in dovi_tool for RPU extraction? If we're starting with a FEL input, can one use either Mode 1 or Mode 2 in your tool to extract the RPU to be used to create a BL+RPU stream/file? In other words, what is the difference between doing a “FEL to MEL” conversion vs “FEL to Profile 8.1”?

**Reply by quietvoid** [reply]
If your final file is only BL+RPU, it's better to use mode 2.
I usually suggest using mode 0 to extract, and then doing further modifications with [the editor](https://github.com/quietvoid/dovi_tool/blob/main/docs/editor.md).
That way you can always come back to the original RPU without having to extract again.

**Reply by markojurkovic** [reply]
Thanks for the comment. I tried extracting the RPU using Mode 1 and then injecting it to a Blu-ray-sourced Base Layer and then muxed it to mp4 using Dolby’s muxer, and the file gets detected as Dolby Vision on my TV, so I guess it worked. Going to use that procedure from now on to create Dolby Vision (MEL) files for playback on the TV, which cannot decode FEL data ostensibly.

**Comment by markojurkovic** [comment]
Hello quietvoid, just out of curiosity, what is the difference between “MEL RPU” vs “Profile 8.1 RPU”?

**Reply by quietvoid** [reply]
MEL is a type of profile 7 RPU.
It's meant to be within an enhancement layer HEVC stream.

Profile 8.1 does not use an enhancement layer, so the RPU is in an HDR10 HEVC stream.

So in a profile 7 RPU, some of the display management flags/configs are different.

**Reply by markojurkovic** [reply]
Thanks! So converting between the different modes of RPU is lossless to the source RPU, correct? The difference between the modes is just in the way the RPU is interleaved with the BL?

**Reply by quietvoid** [reply]
No, not all the modes round trip losslessly. Parameters in the RPU are changed, or even removed completely.

**Reply by markojurkovic** [reply]
Thanks, in the specific case of Profile 7 RPU to Profile 8.1 RPU conversion, is that lossless?

**Reply by quietvoid** [reply]
Only for MEL.
